courses/utils.py

```diff
@@ -0,0 +1,220 @@
+        )
+    else:
+        CourseRunGrade.objects.update_or_create(
+            user=user,
+            course_run=course_run,
+            defaults={
+                "grade": edx_grade.percent,
+                "passed": edx_grade.passed,
+                "letter_grade": edx_grade.letter_grade,
+            },
+        )
+
+
+def sync_courserun_grades(readable_id, email=None, override_grade=None, force=False):
+    """
+    Synchronize grades from OpenEdX to xPRO database
+    """
+    _verify_configuration()
+
+    edx_client = EdxApi(
+        {
+            "access_token": settings.EDX_ACCESS_TOKEN,
+            "api_key": settings.OPENEDX_API_KEY,
+        },
+        settings.OPENEDX_API_BASE_URL,
+    )
+    user = None
+    if email:
+        user = fetch_user(email)
+
+    course_run = _get_course_run(readable_id)
+
+    # Grade can also be 0.0 so `if not grade` does not work here
+    # We only sync grades if the grade is not provided
+    if override_grade is None:
+        # Grade synced from edX for a specific user for the course run
+        if user:
+            edx_grade = edx_client.current_grades.get_student_current_grade(
+                user.username, course_run.courseware_id
+            )
+            _generate_grade(
+                user=user, course_run=course_run, edx_grade=edx_grade, force=force
+            )
+
+        # Grades synced from edX for all enrolled users for the course run
+        else:
+            grades = edx_client.current_grades.get_course_current_grades(
+                course_run.courseware_id
+            )
+            all_grades = list(grades.all_current_grades)
+            for edx_grade in all_grades:
+                try:
+                    user = fetch_user(edx_grade.email)
+                    _generate_grade(
+                        user=user,
+                        course_run=course_run,
+                        edx_grade=edx_grade,
+                        force=force,
+                    )
+                except User.DoesNotExist as e:
+                    logger.warning(str(e))
+    else:
+        # Grade override for all users for the course run. Disallowed.
+        if not user:
+            exception = PermissionDenied(
+                "No user supplied with override grade. Force overwrite of grade is not supported \
+                    for all users. Grade should only be supplied when a specific user is targeted."
+            )
+            logger.error(str(exception))
+            raise exception
+        # Grade override for specific user for the course run
+        else:
+            if (
+                CourseRunGrade.objects.filter(user=user, course_run=course_run).exists()
+                and not force
+            ):
+                logger.warning(
+                    "Grade exists for user: %s and courseware_id: %s. Not overwriting",
+                    user.email,
+                    course_run.courseware_id,
+                )
+            else:
```

Comment:
We can eliminate this `else` too if just `return` in case of `Grade exists for user`

Reply:
Done